facebook/react-native
# Implement a StatusBar frame API

## Motivation

Status Bars are far too variable to hardcode, they are different on a variety of devices and can change at runtime:

* We worked around the Android issues temporarily with StatusBar.currentHeight. However we have not handled ios,
* The In-Call Status Bar makes the real height of the status bar 40px instead of 20px, though it pushes the app's root down by 20px.
* In iOS 11 it's now standard for the status bar height to be 0 in landscape mode. (The iOS simulator in Xcode 9 seems to do this occasionally on iOS before 11 as well)
* Apple has a new iPhone X that has a 44px status bar height instead of 20px, and it doesn't change height when the In-Call Status bar is shown.

The solution is a new api to get the Status Bar's frame (the height of the translucent Status Bar overlap and the y offset of that overlap if an In-Call Status bar is used).

- A frame is an object with the StatusBar's height and y offset:
  - height: Is the height of the StatusBar that would overlap the app if it were translucent.
  - y: Is the y offset of the StatusBar. This is used in the case of the iOS In-Call Status Bar because it doesn't increase the size of the overlap, it just pushes the content area down another 20px.
- `getCurrentFrame` returns a Promise for the current frame
- The `statusBarFrameDidChange` event allows you to listen to changes in the frame.

## Breaking

This changes the unofficial `getHeight` and `statusBarFrame{Did,Will}Change` events in `NativeModules.StatusBarManager`, which are used by the 3rd party `react-native-status-bar-size` library. This will break that library, but the new API should be able to replace it.

Notable differences:

* These frame objects do not return an `x` and `width` like the old ones. x is never not `0` and `width` is just the same as the viewport you can get with `Dimensions`. The Android version would be pointlessly complex if I tried to calculate the width of the viewport just to provide this.
* The `height` this returns is the "overlap height" of the area that you need to compensate for with padding in your nav bars, drawers, etc... instead of the "real status bar height" that was previously returned. However users of that library who really want the "real height" can just do `frame.y + frame.height` to get it.
* I had to drop `statusBarFrameWillChange` because I cannot do the offset calculation in that event.

I've started a discussion in that library's issue tracker to see how these differences affect users of that library:
https://github.com/jgkim/react-native-status-bar-size/issues/19

## Test Plan

* Open the RNTester app
* Open the "`<StatusBar>`" examples
* Scroll down to the bottom to see the status bar's dimensions. Do various things like changing the orientation, toggling the In-Call status bar, and trying it out on different devices.

<details>
<summary><b>Screenshots</b></summary>

![screenshot_1508534406](https://user-images.githubusercontent.com/53399/31842979-992f3608-b5a5-11e7-83c7-d12f061b9ab7.png)
![screenshot_1508534352](https://user-images.githubusercontent.com/53399/31842980-99456842-b5a5-11e7-8088-1da4627d1dd7.png)
![simulator screen shot - iphone x - 2017-10-20 at 14 11 59](https://user-images.githubusercontent.com/53399/31842981-995cecb0-b5a5-11e7-93b4-0a928a1b0bab.png)
![simulator screen shot - iphone 8 - 2017-10-20 at 14 08 46](https://user-images.githubusercontent.com/53399/31842983-99733920-b5a5-11e7-975e-5ae714a5efa4.png)
![simulator screen shot - iphone 8 - 2017-10-20 at 14 08 33](https://user-images.githubusercontent.com/53399/31842984-9989c622-b5a5-11e7-8b9a-f49c734c015a.png)
![simulator screen shot - iphone 8 - 2017-10-20 at 14 08 29](https://user-images.githubusercontent.com/53399/31842985-99a0c246-b5a5-11e7-9bc4-df16f913360d.png)
![simulator screen shot - iphone x - 2017-10-20 at 14 07 20](https://user-images.githubusercontent.com/53399/31842987-99b82f6c-b5a5-11e7-9105-31079a9dd1e1.png)
</details>
